airbyte-workers/src/main/java/io/airbyte/workers/process/AsyncOrchestratorPodProcess.java

```diff
@@ -0,0 +1,282 @@
+ * launching pod and the launched pod to co-exist for the duration of execution for the launched
+ * pod.
+ *
+ * Instead, this process creates the pod and interacts with a document store on cloud storage to
+ * understand the state of the created pod.
+ */
+@Slf4j
+public class AsyncOrchestratorPodProcess implements KubePod {
+
+  public static final String KUBE_POD_INFO = "KUBE_POD_INFO";
+  public static final String NO_OP = "NO_OP";
+
+  private final KubePodInfo kubePodInfo;
+  private final DocumentStoreClient documentStoreClient;
+  private final KubernetesClient kubernetesClient;
+  private final AtomicReference<Optional<Integer>> cachedExitValue;
+
+  public AsyncOrchestratorPodProcess(
+                                     final KubePodInfo kubePodInfo,
+                                     final DocumentStoreClient documentStoreClient,
+                                     final KubernetesClient kubernetesClient) {
+    this.kubePodInfo = kubePodInfo;
+    this.documentStoreClient = documentStoreClient;
+    this.kubernetesClient = kubernetesClient;
+    this.cachedExitValue = new AtomicReference<>(Optional.empty());
+  }
+
+  public Optional<String> getOutput() {
+    return getDocument(AsyncKubePodStatus.SUCCEEDED.name());
+  }
+
+  private int computeExitValue() {
+    final AsyncKubePodStatus docStoreStatus = getDocStoreStatus();
+
+    // trust the doc store if it's in a terminal state
+    if (docStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+      return 1;
+    } else if (docStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+      return 0;
+    }
+
+    final Pod pod = kubernetesClient.pods()
+        .inNamespace(getInfo().namespace())
+        .withName(getInfo().name())
+        .get();
+
+    // Since the pod creation blocks until the pod is created the first time,
+    // if the pod no longer exists (and we don't have a success/fail document)
+    // we must be in a state where it completed and therefore failed
+    if (pod == null) {
+      return 1;
+    }
+
+    // If the pod does exist, it may be in a terminal (error or completed) state.
+    final boolean isTerminal = KubePodProcess.isTerminal(pod);
+
+    if (isTerminal) {
+      // In case the doc store was updated in between when we pulled it and when
+      // we read the status from the Kubernetes API, we need to check the doc store again.
+      final AsyncKubePodStatus secondDocStoreStatus = getDocStoreStatus();
+      if (secondDocStoreStatus.equals(AsyncKubePodStatus.FAILED)) {
+        return 1;
+      } else if (secondDocStoreStatus.equals(AsyncKubePodStatus.SUCCEEDED)) {
+        return 0;
+      } else {
+        // otherwise, the actual pod is terminal when the doc store says it shouldn't be.
+        return 1;
+      }
+    }
+
+    // Otherwise, throw an exception because this is still running.
+    switch (docStoreStatus) {
+      case NOT_STARTED -> throw new IllegalThreadStateException("Pod hasn't started yet.");
```

Comment:
Nit: the exception is a little bit confusing since it is not really an error. Could you add a comment that it will be catch in `hasExited` method?